code/headers/display_adapter.hpp

```diff
@@ -61,13 +89,282 @@ namespace r2d2::display {
          * @param data
          */
         virtual void set_pixels(uint16_t x, uint16_t y, uint16_t width,
-                                uint16_t height, const uint16_t data);
+                                uint16_t height, const uint16_t data) {
+            // set all the pixels to data
+            for (std::size_t t_y = 0; t_y < width; t_y++) {
+                for (std::size_t t_x = 0; t_x < width; t_x++) {
+                    // set the pixel with datas
+                    set_pixel(x + t_x, y + t_y, data);
+                }
+            }
+        }
+
+        /**
+         * @brief Sets character in a single color
+         *
+         * @param x x-coordinate of the character (x=0 is the leftmost collumn)
+         * @param y y-coordinate of the character (y=0 is the highest row)
+         * @param character The un-extended (0-127) ascii value of the character
+         * @param pixel_color The color of the character
+         */
+        virtual void set_character(uint16_t x, uint16_t y, char character,
+                                   uint16_t pixel_color) {
+            // Loops through all rows of the character
+            const hwlib::image &character_image = display_font[character];
+            for (uint16_t image_y = 0; image_y < 8; image_y++) {
+                for (uint16_t image_x = 0; image_x < 8; image_x++) {
+                    // Check if the pixel needs to be drawn.
+                    // If the pixel color is anything other than white, draw it
+                    hwlib::color character_pixel_color =
+                        character_image[hwlib::xy(image_x, image_y)];
+                    if (character_pixel_color != hwlib::white) {
+                        set_pixel(x + image_x, y + image_y, pixel_color);
+                    }
+                }
+            }
+        }
+
+        /**
+         * @brief Sets characters in a single color
+         *
+         * @param x x-coordinate of the first character
+         * @param y y-coordinate of the first character
+         * @param characters Array of characters to draw
+         * @param pixel_color The color of all characters
+         */
+        virtual void set_character(uint16_t x, uint16_t y,
+                                   const char *character,
+                                   uint16_t pixel_color) {
+            std::size_t index = 0;
+            while (character[index] != '\0') {
+                set_character(x, y, character[index], pixel_color);
+
+                // If the cursor is about to go out of bounds, return.
+                if (x + 8 < DisplayScreen::width) {
+                    x += 8;
+                } else {
+                    return;
```

Comment:
at this point this is a design choice which is pointed out on the wiki page i belive it is okay since the PO has no requirements in this regard(atleast as far as i know)

Reply:
Alright, didn't know this was on the wiki.